flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/AsyncSinkBaseBuilder.java

```diff
@@ -37,6 +38,7 @@
         ConcreteBuilderT extends AsyncSinkBaseBuilder<?, ?, ?>> {
 
     private ElementConverter<InputT, RequestEntryT> elementConverter;
+    private FatalExceptionHandler<RequestEntryT> fatalExceptionHandler;
```

Comment:
I have been discussing with @CrynetLogistics about fully encapsulating the `RequestEntryT` to the user code. At the moment it is unnecessarily bubbled up, breaking encapsulation and preventing us to change the backing implementation with breaking user code. This now poses an interesting challenge in terms of implementation with `FatalExceptionHandler`. I would like @CrynetLogistics opinion here on how we can add this without exposing `RequestEntryT`.

Reply:
Thanks, as I commented below, if we think the requestEntries should not be passed to the handler (as we don't believe letting the user access to the failed request entries - in order to log, publish to DLQ etc makes sense) , then I agree, no need to pass these

Reply:
It would be a bigger change, but maybe instead of passing the `RequestEntryT` we could pass the raw objects. Since `RequestEntryT` is destination specific it might not be ideal for forwarding on, and we do not want to require the exception handler to deserialise. Would appreciate input from @CrynetLogistics here .

Reply:
Hi @nirtsruya thanks for your contribution.

I see the following options:

1. Keeping a mapping between `InputT` and `RequestEntryT` in the `SinkWriter` and only exposing the `InputT`.
Advantages:
- Only exposes the `InputT` to the user (would not break encapsulation, underlying sink implementation free to change)
Disadvantages:
- Sink would require ~2x the memory to store the corresponding `InputT`s
- Would complicate `SinkWriter` implementation

2. Return a `toString`-style printout of the failed `RequestEntryT`s - since most people really only want to append a representative list of failed entries in an appropriate format to their message log, we can provide a `shortFailedList` with a few elements (or under 200 chars) and a `fullFailedList`. The sink implementer would be in the best position to determine what is a 'representative list of failed entries in an appropriate format' for their particular set of `RequestEntryT`s.
Advantages:
- Would not expose `RequestEntryT` to the user (would not break encapsulation, underlying sink implementation free to change)
Disadvantages:
- User not free to use the `RequestEntryT`s in whatever way they like - sort/concat/parse&determine error/etc.